File: src/gaddag.py
import pickle
import os
import cProfile

from collections import deque

from letter_mask import *

_CHARS = 'ABCDEFGHIJKLMNOPQRSTUVWXYZ^'


class Arc:
    __slots__ = ['destination_state', 'letter_set']

    def __init__(self, destination_state: State):
        self.destination_state = destination_state
        self.letter_set = 0 #Start with empty set of letters

    # def __str__(self):
    #     return str(self.destination_state) + "Letter set: " + str(self.letter_set) + "\n"

    def next_arc(self, letter):
        return self.destination_state.arcs.get(letter) 
    
class State:
    __slots__ = ['arcs']

    def __init__(self):
        self.arcs = {}

    # def __str__(self):
    #     return "Next arcs: " + str(self.arcs.keys()) + "\n"

class Gaddag:
    def __init__(self, path = 'wordList.txt', cache_path = 'gaddag.pkl'):
        if os.path.exists(cache_path) and False:
            with open(cache_path, 'rb') as f:
                cached = pickle.load(f)
                self.root = cached.root
                self.init = cached.init
                self.size = cached.size


        else:
            self.root = State()
            self.init = Arc(self.root)
            self.init.letter_set = None

            self.size = 1

            with open(path, 'r') as file:
                for line in file:
                    for word in line.strip().split():
                        self.add_word(word.upper())

            # with open(cache_path, 'wb') as f:
            #     pickle.dump(self, f)

        print(f"Gaddag has {self.size} nodes")

    def add_word(self, word: str):

        def _add_arc(st: State, ch: str) -> State:
            if ch not in st.arcs:
                st.arcs[ch] = Arc(State())
                self.size += 1
            return st.arcs[ch].destination_state

        def _add_final_arc(st: State, c1: str, c2: str):
            if c1 not in st.arcs:
                st.arcs[c1] = Arc(State())
                self.size += 1
            st.arcs[c1].letter_set = add_char(st.arcs[c1].letter_set, c2)
            return st.arcs[c1].destination_state

        def _force_arc(st: State, ch: str, force_st: State):
            if ch in st.arcs:
                existing = st.arcs[ch].destination_state
                if existing is not force_st:
                    raise ValueError(f"Conflict: existing destination {existing} is not the forced destination {force_st}")
            else:
                st.arcs[ch] = Arc(force_st)
            return st.arcs[ch]
                
        n = len(word)
        a = word

        current_state = self.root
        for i in range(n - 1, 1, -1):
            current_state = _add_arc(current_state, a[i])
        _add_final_arc(current_state, a[1], a[0])

        current_state = self.root
        for i in range(n - 2, -1, -1):
            current_state = _add_arc(current_state, a[i])
        current_state = _add_final_arc(current_state, '^', a[n - 1])

        first_iteration = True
        for m in range(n - 2, 0, -1):
            force_st = current_state
            current_state = self.root
            for i in range(m - 1, -1, -1):
                current_state = _add_arc(current_state, a[i])
            current_state = _add_arc(current_state, '^')
            arc = _force_arc(current_state, a[m], force_st)

            if first_iteration:
                arc.letter_set = add_char(arc.letter_set, a[n - 1])
                first_iteration = False

    def get_init(self):
        return self.init

   
                          

File: C++/Vision/Gaddag.cpp
#include <fstream>
#include <sstream>
#include <stdexcept>
#include <cctype>
#include "Gaddag.h"

static constexpr char END_MARKER = '^';

Gaddag::Gaddag(const std::string& path) {
    root = makeNode();          // nodes[0] is root; size starts at 1
    init.destination_state = root;

    std::ifstream file(path);
    std::string line;
    while (std::getline(file, line)) {
        std::istringstream iss(line);
        std::string word;
        while (iss >> word) {
            for (char& c : word) c = static_cast<char>(std::toupper(static_cast<unsigned char>(c)));
            add_word(word);
        }
    }
}

Node* Gaddag::makeNode() {
    nodes.push_back(std::make_unique<Node>());
    return nodes.back().get();
}

Node* Gaddag::add_arc(Node* st, char ch) {
    Arc& arc = st->get_or_create_arc(ch);
    if (arc.destination_state == nullptr) {
        arc.destination_state = makeNode();
        ++size;
    }
    return arc.destination_state;
}

Node* Gaddag::add_final_arc(Node* st, char c1, char c2) {
    Arc& arc = st->get_or_create_arc(c1);
    if (arc.destination_state == nullptr) {
        arc.destination_state = makeNode();
        ++size;
    }
    arc.letter_mask |= (1u << char_to_index(c2));
    return arc.destination_state;
}

Arc& Gaddag::force_arc(Node* st, char ch, Node* force_st) {
    Arc& arc = st->get_or_create_arc(ch);
    if (arc.destination_state != nullptr && arc.destination_state != force_st) {
        throw std::runtime_error("Conflict: existing destination is not the forced destination");
    }
    arc.destination_state = force_st;
    return arc;
}

void Gaddag::add_word(const std::string& word) {
    const int n = static_cast<int>(word.size());

    Node* current_state = root;
    for (int i = n - 1; i >= 2; --i)            // range(n-1, 1, -1)
        current_state = add_arc(current_state, word[i]);
    add_final_arc(current_state, word[1], word[0]);

    current_state = root;
    for (int i = n - 2; i >= 0; --i)            // range(n-2, -1, -1)
        current_state = add_arc(current_state, word[i]);
    current_state = add_final_arc(current_state, END_MARKER, word[n - 1]);

    bool first_iteration = true;
    for (int m = n - 2; m >= 1; --m) {          // range(n-2, 0, -1)
        Node* force_st = current_state;
        current_state = root;
        for (int i = m - 1; i >= 0; --i)
            current_state = add_arc(current_state, word[i]);
        current_state = add_arc(current_state, END_MARKER);
        Arc& arc = force_arc(current_state, word[m], force_st);

        if (first_iteration) {
            arc.letter_mask |= (1u << char_to_index(word[n - 1]));
            first_iteration = false;
        }
    }
}
